lib/libutee/tee_api_operations.c

```diff
@@ -464,97 +476,119 @@ TEE_Result TEE_SetOperationKey2(TEE_OperationHandle operation,
 {
 	TEE_Result res;
 	uint32_t key_size = 0;
+	TEE_ObjectInfo key_info1;
+	TEE_ObjectInfo key_info2;
 
-	if (operation == TEE_HANDLE_NULL)
-		TEE_Panic(0);
-
-	/* Two keys not expected */
-	if ((operation->info.handleState & TEE_HANDLE_FLAG_EXPECT_TWO_KEYS) ==
-	    0)
-		TEE_Panic(0);
+	/* Operation is not a valid handle */
+	if (operation == TEE_HANDLE_NULL) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
-	/* Either both keys are NULL or both are not NULL */
+	/*
+	 * Key1/Key2 and/or are not initialized
+	 * Either both keys are NULL or both are not NULL
+	 */
 	if ((key1 == TEE_HANDLE_NULL || key2 == TEE_HANDLE_NULL) &&
-	    key1 != key2)
-		TEE_Panic(0);
+	    key1 != key2) {
+		/* Operation key1 cleared */
+		TEE_ResetTransientObject(operation->key1);
+		/* Operation key2 cleared */
+		TEE_ResetTransientObject(operation->key2);
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
-	if (key1 != TEE_HANDLE_NULL) {
-		TEE_ObjectInfo key_info1;
-		TEE_ObjectInfo key_info2;
+	/* No key for digest operation */
+	if (operation->info.operationClass == TEE_OPERATION_DIGEST) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
-		res = TEE_GetObjectInfo1(key1, &key_info1);
-		if (res != TEE_SUCCESS)
-			goto err;
+	/* Two keys flag expected (TEE_ALG_AES_XTS only) */
+	if ((operation->info.handleState & TEE_HANDLE_FLAG_EXPECT_TWO_KEYS) ==
+	    0) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
 
-		/* Supplied key has to meet required usage */
-		if ((key_info1.objectUsage & operation->info.
-		     requiredKeyUsage) != operation->info.requiredKeyUsage) {
-			TEE_Panic(0);
-		}
+	res = TEE_GetObjectInfo1(key1, &key_info1);
```

Comment:
It looks like when key1 key2 are both TEE_HANDLE_NULL, it leads to panic(0) at the end of the function. But spec specifies one of panic reasons as follows.
 "key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object."
To me, it looks like the function shouldn't panic when both key1 and key2 are TEE_HANDLE_NULL.